#1 Pick one of the datasets from the ChatBot session(s) of the TUT demo 
(or from your own ChatBot session if you wish) and use the code produced through the 
ChatBot interactions to import the data and confirm that the dataset has missing values.

In [16]:
# feel free to just use the following if you prefer...
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

row_n           0
id              1
name            0
gender          0
species         0
birthday        0
personality     0
song           11
phrase          0
full_id         0
url             0
dtype: int64

#2 Start a new ChatBot session with an initial prompt introducing the dataset 
you're using and request help to determine how many columns and rows of data a pandas DataFrame has, and then
use code provided in your ChatBot session to print out the number of rows and columns of the dataset; and,
write your own general definitions of the meaning of "observations" and "variables" based on asking the ChatBot 
to explain these terms in the context of your dataset.

In [17]:
import pandas as pd

# Load your dataset 
df = pd.read_csv('https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv')

# Get the number of rows and columns
rows, columns = df.shape

print(f'The dataset has {rows} rows and {columns} columns.')


The dataset has 391 rows and 11 columns.


In [ ]:
# Observations: This is a row in the dataset. In this case it would be just one villager we study and all the information (ex. name, gender, species, birthday, etc.) we have on him or her. 
# Variables: This is the type of information we have on observations like their name, gender, species, birthday, etc. In the context of a dataset, it is the columns.

#3 Ask the ChatBot how you can provide simple summaries of the columns in the dataset and use the suggested code to 
provide these summaries for your dataset.

In [26]:
# Simple summary of numerical columns
summary = df.describe()
print(summary)

print(" \n ")

# General info of the DataFrame (including non-numeric columns)
general_info = df.info()

print(" \n ")

# Count of uniques values in a column
value = df['species'].value_counts()
print(value)

print(" \n ")

# Summary of categorical columns
categorical_summary = df.describe(include=['object'])
print(categorical_summary)


            row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000
 
 
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 391 entries, 0 to 390
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   row_n        391 non-null    int64 
 1   id           390 non-null    object
 2   name         391 non-null    object
 3   gender       391 non-null    object
 4   species      391 non-null    object
 5   birthday     391 non-null    object
 6   personality  391 non-null    object
 7   song         380 non-null    object
 8   phrase       391 non-null    object
 9   full_id      391 non-null    object
 10  url          391 non-null    object
dtypes: int64(1), object(10)
memory usage: 33.7+ KB
 
 
species
cat          23
rabbit       20
frog         18
squirrel     18
duck         17
dog          16
cub          16
pig          15
bear

In [ ]:
# df.describe(): Summarizes all the numeric columns with  metrics like count, mean, std, min, 25%, 50%, 75%, and max.

# df.info(): Displays information about the dataset such as data types of each column, the number of non-null entries, and memory usage.

# df['column'].value_counts() counts the occurrences of each unique value in a specified DataFrame column and returns them in descending order.

# df.describe(include=['object']): Summarizes non-numeric columns with details like count, unique, top (most frequent category), and freq (frequency of the top category).

#4 If the dataset you're using has (a) non-numeric variables and (b) missing values in numeric variables, explain 
(perhaps using help from a ChatBot if needed) the discrepancies between size of the dataset given by df.shape and what 
is reported by df.describe() with respect to (a) the number of columns it analyzes and (b) the values it reports in 
the "count" column.

In [30]:
Number of columns analysed:
# df.shape: Shows the total number columns in the dataset regardless of data types.
# df.describe: Summarizes only numerical columns so non-numerical ones are excluded.

Values reported:
# df.shape: Shows the total number of rows.
# df.describe: Reports the count of non-NaN values for each numeric column. If there are missing values, the count will be less than the total number of rows.

#5 Use your ChatBot session to help understand the difference between an "attribute", such as df.shape which does not end with () and a "method", such as df.describe() which does end with ().

Provide your own paraphrasing summarization of that difference. 

In [3]:
Attribute:
# An attribute is a property of an object that has stored data about such object. An attribute such as df.shape will simply let one access the stored information. It does not need parentheses because it is just returning values and not taking any action. 
    
Method:
# A method is a function associated with an object. A method such as df.describe, when called, will perform an action; meaning it will compute or modify the data. Methods require parentheses so arguments (information needed for the action to happen) can be placed inside.

#6 The df.describe() method provides the 'count', 'mean', 'std', 'min', '25%', '50%', '75%', and 'max' summary statistics for each variable it analyzes. Give the definitions (perhaps using help from the ChatBot if needed) of each of these summary statistics.

In [ ]:
Count:
#The number of non-NaN variables in the column. 
    
Mean:
#The average of the values(non-NaN) in the column. It is the sum of all the values(non-NaN) divided by the number of values(non-NaN). 
    
Std:
#Standard deviation. A measurement of the amount of variation in the values of the column. It tells you the average of how far each value lies from the mean. A low standard deviation means that the values lie close to the mean.
    
Min:
#The smallest value in the column

First Quartile:
#The value where 25% of the data lies below
    
Median Quartile:
#The value where 50% of the data lies below

Third Quartile:
#The value where 75% of the data lies below

Max:
#The largest value in the column


#7 Missing data can be considered "across rows" or "down columns". Consider how df.dropna() or del df['col'] should be applied to most efficiently use the available non-missing data in your dataset and briefly answer the following questions in your own words

       Provide an example of a "use case" in which using df.dropna() might be peferred over using del df['col']
   
       Provide an example of "the opposite use case" in which using del df['col'] might be preferred over using df.dropna()
   
       Discuss why applying del df['col'] before df.dropna() when both are used together could be important
   
       Remove all missing data from one of the datasets you're considering using some combination of del df['col'] and/or 
       df.dropna() and give a justification for your approach, including a "before and after" report of the results of your 
       approach for your dataset.

In [ ]:
# Use case of df.dropna() over del df['col']
# df.dropna() allows you to drop rows with any missing values. This could be useful when a company analyses survey results where some people left a field blank. Instead of deleting the entire column where information is needed, it may be better to just drop the incomplete responses.
df_clean = df.dropna()
df_clean

# Use case of del df['col'] over df.dropna()
# del df['col'] allows you to delete any column you specify in the square brackets. This could be useful when a company analyses survey results and see many missing values in non-critical fields such as "middle_name". It may be better to just delete that data field instead of losing responses.
del df['column_name']

# Discuss why applying del df['col'] before df.dropna() when both are used together could be important
# Since df.dropna() will remove any rows with missing data, it is important to make sure the rows with values missing only in the non-critical columns don't get deleted.
del df['irrelevant_col']
df_clean = df.dropna()
df_clean


In [13]:
# For question 4 I will load a new dataset with more missing values
import pandas as pd
link="https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv" 
mn=pd.read_csv(link)
mn

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [14]:
# Finding the missing values
mn.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [15]:
# Step 1: Delete the 'deck' column since it has too many missing values
del mn['deck']


# Step 2: Drop rows that contain any missing values in remaining columns
mn_clean = mn.dropna()

# Report the result
print(f"Before: {mn.shape}")  
print(f"After: {mn_clean.shape}") 

Before: (891, 14)
After: (712, 14)


Before: There are 819 rows and 15 columns. Column 'age' has around 20% missing values while 'deck' has around 77% missing values. There are also rows missing values in the 'embarked' and 'embarked_town columns'.

After: The dataset now has 1 less column and 179 fewer rows. This is because I removed one column 'deck' and removed 177 rows with no value in 'age' and 2 rows with no value in 'embarked' and 'embarked_town'. 

Justification: Like mentioned in the previous question, removing insignificant columns first will preserve the number of people recorded in this dataset. 

#8 Give brief explanations in your own words for any requested answers to the questions below:
    
    1. Use your ChatBot session to understand what df.groupby("col1")["col2"].describe() does and then demonstrate and 
    explain this using a different example from the "titanic" data set other than what the ChatBot automatically provide for 
    you.
    
    2. Assuming you've not yet removed missing values in the manner of question "7" above, df.describe() would have 
    different values in the count value for different data columns depending on the missingness present in the original 
    data. Why do these capture something fundamentally different from the values in the count that result from doing 
    something like df.groupby("col1")["col2"].describe()?
    
    3. Intentionally introduce the following errors into your code and report your opinion as to whether it's easier 
    to (a) work in a ChatBot session to fix the errors, or (b) use google to search for and fix errors: first share 
    the errors you get in the ChatBot session and see if you can work with ChatBot to troubleshoot and fix the coding 
    errors, and then see if you think a google search for the error provides the necessary toubleshooting help more 
    quickly than ChatGPT.
    
    a) Forget to include import pandas as pd in your code
    b) Mistype "titanic.csv" as "titanics.csv"
    c) Try to use a dataframe before it's been assigned into the variable
    d) Forget one of the parentheses somewhere the code
    e) Mistype one of the names of the chained functions with the code
    f) Use a column name that's not in your data for the groupby and column selection
    g) Forget to put the column name as a string in quotes for the groupby and column selection

In [24]:
# Question 1
import pandas as pd

# Load the Titanic dataset
file = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
tt = pd.read_csv(file)

# Group by 'class' and summarize 'fare' 
grouped_summary = tt.groupby("class")["fare"].describe()
grouped_summary

# df.groupby("class"): This groups the dataset by passenger class (1st, 2nd, and 3rd class).
# ["fare"]: This gives the instruction to analyze the fare column within each passenger class.
# .describe(): For each class, this will give us a summary of the fare statistics in that class.

count       mean        std  min       25%      50%   75%       max
class                                                                      
First   216.0  84.154687  78.380373  0.0  30.92395  60.2875  93.5  512.3292
Second  184.0  20.662183  13.417399  0.0  13.00000  14.2500  26.0   73.5000
Third   491.0  13.675550  11.778142  0.0   7.75000   8.0500  15.5   69.5500

In [ ]:
# Question 2:
# For df.describe(), count considers all the non-missing values in the dataset.
# For df.groupby("col1")["col2"].describe(), count considers the non-missing values for col2 for every group defined in col1.

In [ ]:
# Question 3:
# Troubleshooting with a ChatBot is better than using Google because the ChatBot offers an interactive experience that offers immediate feedback.
# Troubleshooting with a ChatBot is better than using Google because the ChatBot can analyse specific situations unlike Google who only offers pages containing information about a broad range of topics.


9. Have you reviewed the course wiki-textbook and interacted with a ChatBot (or, if that wasn't sufficient, real people in the course piazza discussion board or TA office hours) to help you understand all the material in the tutorial and lecture that you didn't quite follow when you first saw it?

In [16]:
print('yes')

yes
